finetuner/tuner/pytorch/losses.py

```diff
@@ -1,143 +1,135 @@
 
-    This loss is also known as contrastive loss.
+    The loss for a pair of objects equals ::
 
-    The loss being optimized equals::
+        is_sim * dist + (1 - is_sim) * max(0, margin - dist)
 
-        [is_sim * dist + (1 - is_sim) * max(margin - dist, 0)]^2
+    where ``is_sim`` equals 1 if the two objects are similar, and 0 if they are not
+    similar. The ``dist`` refers to the distance between the two objects, and ``margin``
+    is a number to help bound the loss for dissimilar objects.
 
-    where ``target`` should equal 1 when both objects belong to the same class,
-    and 0 otheriwse. The ``dist`` is the euclidean distance between the embeddings of
-    the objects, and ``margin`` is some number, used here to ensure better stability
-    of training.
-
-    The final loss is the average over losses for all pairs of objects in the batch.
+    The final loss is the average over losses for all pairs given by the indices.
     """
 
-    arity = 2
+    def __init__(self, distance: str = "cosine", margin: float = 1.0):
+        """Initialize the loss instance
 
-    def __init__(self, margin: float = 1.0):
+        :param distance: The type of distance to use, avalilable options are
+            ``"cosine"``, ``"euclidean"`` and ``"sqeuclidean"``
+        :param margin: The margin to use in loss calculation
+        """
         super().__init__()
+        self.distance = distance
         self.margin = margin
 
     def forward(
-        self, embeddings: List[torch.Tensor], target: torch.Tensor
+        self,
+        embeddings: torch.Tensor,
+        indices: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
     ) -> torch.Tensor:
-        """Compute the loss.
+        """Compute the loss
 
-        :param inputs: Should be a list or a tuple containing three tensors:
-            - ``[N, D]`` tensor of embeddings of the first objects of the pair
-            - ``[N, D]`` tensor of embeddings of the second objects of the pair
-        :param target: A ``[N, ]`` tensor of target values
+        :param embeddings: An ``[N, d]`` tensor of embeddings
+        :param indices: A list of tuple indices and target, where each element in the
+            list contains three elements: the indices of the two objects in the pair,
+            and their similarity (which equals 1 if they are similar, and 0 if they
+            are dissimilar)
         """
-        l_emb, r_emb = embeddings
-        eucl_dist = F.pairwise_distance(l_emb, r_emb, p=2)
-        is_similar = (target > 0).to(torch.float32)
+        ind_one, ind_two, target = indices
+        dist_matrix = get_distance(embeddings, self.distance)
+        dists = dist_matrix[ind_one, ind_two]
+        target = target.to(torch.float32)
 
-        loss = 0.5 * torch.square(
-            is_similar * eucl_dist + (1 - is_similar) * F.relu(self.margin - eucl_dist)
-        )
+        loss = target * dists + (1 - target) * F.relu(self.margin - dists)
```

Comment:
Why do not use the ContrastiveLoss implementation from `pytorch`?

Reply:
Where do you see ContrastiveLoss in Pytorch?